### ISSUE TITLE
Exception in Destructor

### ISSUE THREAD

**Issue by DasCapschen** [psycopg/psycopg#591]
When a Postgres Connection is deleted, it encounters an Exception in its Constructor, which is fortunately ignored.   
Yet it is a little concerning, that the destructor does not finish successfully.

Code is approximately:
```py
connection = create_postgres_connection()
db_migrator = Migrator(connection=connection)
db_migrator.migrate()
connection.close()
# script ends and destructor is called
```

The error message is:
```
Exception ignored in: <function BaseConnection.__del__ at 0x7f6c249bdbc0>
Traceback (most recent call last):
  File "/opt/maddox/.venv/lib/python3.11/site-packages/psycopg/connection.py", line 157, in __del__
  File "/opt/maddox/.venv/lib/python3.11/site-packages/psycopg/connection.py", line 164, in __repr__
AttributeError: 'NoneType' object has no attribute 'connection_summary'
```

Versions are:
- Python 3.11.4
- Psycopg 3.1.9
- libpg 15.3 (Debian)
- Postgres 15.2 (Alpine Docker Container)

**Comment by dvarrazzo** [comment]
It seems a problem similar to #198.

The `connection_summary` function is pretty complex and I wouldn't try to
create all the closures to make it work at destruction time because it's
untestable. If any I'd try to make repr work, in a degenerate way, on
AttributeError.

**Comment by DasCapschen** [comment]
But the problem does not seem to be in the `connection_summary` function. Instead it already fails when trying to call the connection_summary function. Whatever Object that function is called on, it was already destroyed and does not exist anymore.

**Comment by dvarrazzo** [comment]
Yes, what I meant is that, when the problem was encountered the first time,
for instance with an attribute error on `os.getpid`, the solution was "just
store the `getpid` function in a local var".

This works for simple atomic functions such as getpid, implemented in c and
executed in a single python opcode, but I doubt it would work for
connection_summary: it is likely, and not reliably testable, that an
attribute error might get raised in its body because of some module already
gc'ed.

**Comment by DasCapschen** [comment]
Ah, now I understand what you mean. Seems like it is a Python bug then, if it GC's things that are still used in destructors...
Well okay... guess we'll just keep ignoring this then 😅

**Comment by dvarrazzo** [comment]
A fix for this error will be released in 3.1.10.

**Comment by DasCapschen** [comment]
Wow thank you! I will test this as soon as the release is out :)
